lib/pmdtester/builders/project_hasher.rb

```diff
@@ -117,15 +117,24 @@ def make_violation_hash(file_ref, violation, is_diff = TRUE)
         'l' => violation.line,
         'f' => file_ref,
         'r' => violation.rule_name,
-        'm' => is_diff && violation.changed? ? diff_fragments(violation) : violation.message
+        'm' => create_violation_message(violation, is_diff)
       }
       h['ol'] = violation.old_line if is_diff && violation.changed? && violation.line != violation.old_line
       h
     end
 
-    def diff_fragments(violation)
-      diff = Differ.diff_by_word(violation.message, violation.old_message)
+    def create_violation_message(violation, is_diff)
+      return escape_html(violation.message) unless is_diff
+
+      diff = Differ.diff_by_word(escape_html(violation.message),
+                                 escape_html(violation.old_message))
```

Comment:
thinking out loud here, but is it ok to diff the html? or should we escape the raw diff? I can't think of a reason why it wouldn't be equivalent, but I'm not sure if it would hold for all scenarios…

Reply:
The result of Differ.diff_by_word is itself already html - so we need to escape it before, e.g. it produces something like this:

```html
The method 'triggerHandler(Type, Boolean, Boolean, Boolean, Boolean, Boolean, Boolean, List&lt;SObject&gt;, Map&lt;Id<del class="differ">,SObject&gt;)' has a cognitive complexity of 25</del>, <ins class="differ">SObject&gt;)' has a cognitive complexity of 25, </ins>current threshold is 15
```

It is adding `<del>` and `<ins>` tags...